src/java.base/share/classes/java/lang/String.java

```diff
@@ -2429,10 +2429,65 @@ public int indexOf(int ch) {
+     *
+     * @apiNote
+     * An invocation of this method returns -1 when {@code fromIndex} happens
+     * to be too large. The result is thus indistinguishable from a genuine
+     * absence of {@code ch} in the string.
+     * If stricter behavior is needed, {@link #indexOf(int, int, int)}
+     * should be considered instead. For example,
+     * {@code this.indexOf(ch, fromIndex, this.length())} would throw if
+     * {@code fromIndex} were negative or larger than the string length.
      */
     public int indexOf(int ch, int fromIndex) {
-        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex)
-                          : StringUTF16.indexOf(value, ch, fromIndex);
+        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex, value.length)
+                : StringUTF16.indexOf(value, ch, fromIndex, value.length >> 1);
+    }
+
+    /**
+     * Returns the index within this string of the first occurrence of the
+     * specified character, starting the search at {@code fromIndex} and
+     * stopping before {@code toIndex}.
+     *
+     * <p>If a character with value {@code ch} occurs in the
+     * character sequence represented by this {@code String}
+     * object at an index no smaller than {@code fromIndex} but smaller than
+     * {@code toIndex}, then
+     * the index of the first such occurrence is returned. For values
+     * of {@code ch} in the range from 0 to 0xFFFF (inclusive),
+     * this is the smallest value <i>k</i> such that:
+     * <blockquote><pre>
+     * (this.charAt(<i>k</i>) == ch) &amp;&amp; (fromIndex &lt;= <i>k</i> &lt; toIndex)
+     * </pre></blockquote>
+     * is true. For other values of {@code ch}, it is the
+     * smallest value <i>k</i> such that:
+     * <blockquote><pre>
+     * (this.codePointAt(<i>k</i>) == ch) &amp;&amp; (fromIndex &lt;= <i>k</i> &lt; toIndex)
+     * </pre></blockquote>
+     * is true. In either case, if no such character occurs in this
+     * string at or after position {@code fromIndex} and before position
+     * {@code toIndex}, then {@code -1} is returned.
+     *
+     * <p>All indices are specified in {@code char} values
+     * (Unicode code units).
+     *
+     * @param   ch          a character (Unicode code point).
+     * @param   fromIndex   the index to start the search from (included).
+     * @param   toIndex     the index to stop the search at (excluded).
+     * @return  the index of the first occurrence of the character in the
+     *          character sequence represented by this object that is greater
+     *          than or equal to {@code fromIndex} and less than {@code toIndex},
+     *          or {@code -1} if the character does not occur.
+     * @throws  StringIndexOutOfBoundsException if {@code fromIndex}
+     *          is negative, or {@code toIndex} is larger than the length of
+     *          this {@code String} object, or {@code fromIndex} is larger than
+     *          {@code toIndex}.
```

Comment:
Thanks for the update, I think you've got to a good place.

Reply:
@stuart-marks was quite convincing ;-)